### PR TITLE
Flash read

### PR DESCRIPTION
**Pull request by sjg20**
Little enhancement to 'flash read'

### CODE REVIEW

File: include/util.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Copyright 2020 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_MATH_H_
+#define ZEPHYR_INCLUDE_MATH_H_
+
+/** Return the maximum of a and b */
+#define max(a, b)	\
```
Comment (nashif) [inline include/util.h:11]: did you check MIN and MAX in include/sys/util.h?

Comment (sglass68) [reply include/util.h:11]: :-) OK I need to check my ctags settings. I thought it was odd.

File: drivers/flash/flash_shell.c [hunk 2]
```diff
@@ -153,12 +156,14 @@ static int cmd_read(const struct shell *shell, size_t argc, char *argv[])
...
-	while (cnt--) {
-		u32_t data;
+	cnt *= sizeof(u32_t);
+	for (upto = 0; upto < cnt; upto += todo) {
+		u8_t data[SHELL_HEXDUMP_BYTES_IN_LINE];
 
-		flash_read(flash_dev, addr, &data, sizeof(data));
-		shell_print(shell, "0x%08x ", data);
-		addr += sizeof(data);
+		todo = MIN(cnt, SHELL_HEXDUMP_BYTES_IN_LINE);
```
Comment (nordic-krch) [inline drivers/flash/flash_shell.c:163]: i think it should be `cnt - upto`

**Review by nordic-krch — DISMISSED** [review]

**Comment by sglass68** [comment]
Fixed conflict

File: drivers/flash/flash_shell.c (hunk 2 above)
Comment (sglass68) [reply drivers/flash/flash_shell.c]: oops missed this. Fixed now.

What is the best way to write a test for this? I could use the dummy shell backend perhaps?

Comment (nordic-krch) [reply drivers/flash/flash_shell.c]: not sure how flash works on qemu target and if that can be used for testing. @nvlsianpu ?

Comment (sglass68) [reply drivers/flash/flash_shell.c]: Eek I don't want to use qemu. I just wondered if there was a native/posix driver.

Comment (nvlsianpu) [reply drivers/flash/flash_shell.c]: It does work on qemu an posix using flash simulator driver.

Comment (sglass68) [reply drivers/flash/flash_shell.c]: Qemu is a large overhead for testing a simple thing. It looks like I can use posix for this, although I do see what are apparently bugs in the driver.

https://github.com/zephyrproject-rtos/zephyr/issues/25109

Comment (sglass68) [reply drivers/flash/flash_shell.c]: I found a way to do a native test

**Comment by nordic-krch** [comment]
thanks for the tests. I would like to approve it. You just need to figure out what is wrong with CI. Many warning.

**Comment by sglass68** [comment]
> thanks for the tests. I would like to approve it. You just need to figure out what is wrong with CI. Many warning.

Yes, it needed a board filter. I figured out how to get the logs and how to run locally. However it takes 20 minutes which seems like a long time.

File: drivers/flash/flash_simulator.c [hunk 3]
```diff
@@ -397,10 +397,8 @@ static int flash_mock_init(struct device *dev)
 		return -EIO;
 	}
 
-	if (f_stat.st_size == 0) {
-		/* erase the memory unit by pulling all bits to one */
-		(void)memset(mock_flash, 0xff, FLASH_SIMULATOR_FLASH_SIZE);
-	}
+	/* erase the memory unit by pulling all bits to one */
```
Comment (vanwinkeljan) [inline drivers/flash/flash_simulator.c:400]: This breaks persistency of the flash device, the flash should only be set to all 0xFF if not already exists.

**Review by vanwinkeljan — DISMISSED** [review]
flash persistency is a feature of native posix and flash device should not be erased every time

File: drivers/flash/flash_simulator.c (hunk 3 above)
Comment (sglass68) [reply drivers/flash/flash_simulator.c]: OK I dropped that commit

**Review by nashif — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Comment by sjg20** [comment]
I seem to need more than two approvers for this one. Any takers?

**Review by MaureenHelm — APPROVED** [review]
@jukka why did you cancel the shippable run?

**Comment by carlescufi** [comment]
@sjg20 please rebase

**Pull request merged**